### ISSUE TITLE
Multipath support in demo client

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#1134]
Just enabling the multipath option does not cause the client to actually use multiple paths. Logically, it should. But we need to design that well.

The first idea would be to add an option to specify an additional path. This is simple to design, but it requires in practice adding support for multiple sockets, which implies a pretty large change in the socket loop code. Also, it is difficult to use, because the user has to find out which of the local addresses the client will chose by default, and thus which one should be the other address.

Another design would be to do that automatically inside the client if multipath is enabled. For example, list the IP addresses available on the client using a system API, then automatically create paths for those that are not in use. The new path could in theory use the same port and UDP socket as the other paths.

**Comment by ytxing** [comment]
Hi. Does it mean that multipath function is now available? I notice that there is a simple-multipath branch updated recently. Can I use that for some testing and is there any advice you may give?

Nice day :)

**Comment by huitema** [comment]
The multipath support is almost there. In `picoquicdemo`, the "-M" option will enable its negotiation. It is still very rough, and the "simple-multipath" branch is work in progress to fix some of the issues. The server code will work, but the option will only be activated if the picoquicdemo client starts a new path. That requires creating several sockets bound to different local IP addresses, and I have not gotten to do that yet. Of course, PR would be welcome.

**Comment by huitema** [comment]
@ytxing are you still interested? I am thinking of adding an option to specify an "alternate source address" to the demo program. If the option is specified, the program would start a second path from that address immediately after completing the handshake.

**Comment by ytxing** [comment]
I am afraid that I could not help you for now since I am working on some other issues. But when I am free I am willing to help.😀-------- 原始邮件 --------发件人： Christian Huitema ***@***.***>日期： 2021年11月15日周一 03:07收件人： private-octopus/picoquic ***@***.***>抄送： ytxing ***@***.***>, Mention ***@***.***>主    题： Re: [private-octopus/picoquic] Multipath support in demo client (#1134)
@ytxing are you still interested? I am thinking of adding an option to specify an "alternate source address" to the demo program. If the option is specified, the program would start a second path from that address immediately after completing the handshake.

**Comment by clarkzjw** [comment]
Hi @huitema, I'm interested in the current status of the multipath implementation of picoquic. 
I have already set up and ran a single path experiment, e.g., 
```
./picoquicdemo -c ./ca/ca-cert.pem -k ./ca/server-key.pem -p 4433 -w ./static/
./picoquicdemo -o /tmp localhost 4433
```

I saw a `-M` option in `picoquicdemo` and based on previous discussions, the demo client hasn't fully implemented the multipath features. And I saw [another discussion](https://github.com/private-octopus/picoquic/issues/1344#issuecomment-1086252129). So if I would like to test multipath features, based on my understanding:

1. start `picoquicdemo`, and add `-M` option to announce multipath support.
2. on the client-side, I have to implement it on my own by referencing the examples in `mutipath_test.c`. 

Is that right? Thanks.

**Comment by huitema** [comment]
1) Yes, the "-M" option will enable multipath support on the demo server. It also enables support on the client side, but we do have client side code to open the second path.

2) On the client side, we need a way to add an IP alternate address. Or maybe two addresses, in case one of them would be chosen for the default path. Need to find an available letter and add support for this CLI option. i would keep it local to the client program rather than trying to force it in "config.c".

3) On the client side, we need to add code in the "client loop" to probe the alternate address and port if it is present. That means adding code executed after the handshake, sothe client call picoquic_probe_new_path(), as inside the function multipath_test_one, lines 538-541. 

@clarkzjw if you start a PR, I will review and help.

**Comment by clarkzjw** [comment]
@huitema Thanks for the reply. So basically we have everything we may need (at least on the MPQUIC side), and we just need more work on the `picoquicdemo`.
I encountered MPQUIC in my work, and I may refer to `picoquic` to back my implementation. That being said, I'll try to figure out how it works and reproduce a minimal POC first.

**Comment by clarkzjw** [comment]
Hi @huitema, after digging the code for a while, I haven't figured out several things yet:

1. In [`picoquic_probe_new_path_ex`](https://github.com/private-octopus/picoquic/blob/master/picoquic/quicctx.c#L1896), it calls [`picoquic_assign_peer_cnxid_to_path`](https://github.com/private-octopus/picoquic/blob/master/picoquic/quicctx.c#L1927) after `picoquic_create_path` succeeds. 

However, in `picoquic_assign_peer_cnxid_to_path`, it only assigns `available_cnxid` to `cnx->path[path_id]->p_remote_cnxid`.(https://github.com/private-octopus/picoquic/blob/master/picoquic/quicctx.c#L1884-L1887) What I cannot figure out right now is that when `cnx->path[path_id]->p_local_cnxid` is assigned for the newly created path?

2. Besides, I also noticed in the comments that https://github.com/private-octopus/picoquic/blob/master/picoquic/picoquic_internal.h#L931-L933
```
* As soon as a path is validated, it moves to position 0. The old path[0] moves to the
* last position, and is marked as deprecated. After about 1 RTT, the path resource
* are freed. (TODO: once we actually support multipath, change that behavior.)
```
Is it still the case right now? If the old path is marked as deprecated, how can the application make use of multiple paths?

3. At first, I thought there would be actual `probing` inside `picoquic_probe_new_path_ex`, i.e., try to send packets between client and server on the new path to verify the path works. But it seems that `picoquic_probe_new_path_ex` actually just compares the new path with existing paths managed by picoquic, and if it doesn't exist yet, it creates the path object and store in memory. Is this understanding correct?

Thanks.

**Comment by huitema** [comment]
1) The local cnxid is picked by the peer, when responding to the probe, out of the list provided in "new connection ID" frames. When the 1RTT packet arrives, the code reads the incoming CNXID and assigns it to the new path.

2) the comments predate multipath. They should probably be updated. If multipath is enabled, the behavior is different. The old path is not deprecated, and the path order is not changed.

3) there is a mini state machine for each path. `picoquic_probe_new_path_ex` configures the state of the new path so a probe will be sent as soon as possible.

**Comment by clarkzjw** [comment]
Following some discussion in #1366, I'm not sure whether I understand the expected behaviour of multipath in picoquic.

Let's consider the simplest scenario:

```
eth0:10.0.1.1  -------link 1---------  eth0:10.0.1.2
    Host1                                  Host2
eth1:10.0.2.1  -------link 2---------  eth1:10.0.2.2                         
```
In the above two paths point to point connection, take `Host1` as the `server`, and `Host2` as the `client`. 

1. `picoquicdemo` starts on `Host1` and `Host2` with `-M` enabled. The server address and port `10.0.1.1:4433` is set as the input parameter when the client starts. Assuming initial packets are sent from `10.0.1.2` to `10.0.1.1` between the client and server since it's the default route. 
2. After the initial negotiation finishes, the client calls `picoquic_probe_new_path`, registering (from: `10.0.1.1`, to: `10.0.2.2`) as a new path. (Because `10.0.1.1` is the server address set in step 1, and `10.0.2.2` is the alternative IP for `Host2`)

Then, how do the following packets work? When and how would link2 be utilized to transmit data alongside link1? Is that within the scope of picoquic multipath design or should the upper application layer implement this on their own?

**Comment by huitema** [comment]
The scheduling of packets is defined in the function `picoquic_select_next_path_mp` (sender.c, lines 4335-4493). Each time picoquic is polled, the quic context selects the next connection ready and polls it. Then, the sender code for that connection selects a path on which the next packet shall be sent. If multipath is selected, this done by calling `picoquic_select_next_path_mp`.

There is a state associated to each path:

* if a new path is created, the first packet sent on that path will be a "challenge" packet. This is sent by first selecting that path as the next path on which to send (because it needs to send a probe), and then in "prepare_packet_ready" by calling `picoquic_format_path_challenge_frame` before sending the packet.
* if a challenge is received, the code will send a challenge response at the first occasion.
* if a valid challenge response has been received, the client will start sending data or ACKs on that path. But the server will wait for the path to be "validated".
* if the server receives a non-path-probing frame (data or ack) on the path (after a challenge response) the path is validated and the server start sending data one the path.

How much data depends on the state of congestion control for the path.

**Comment by clarkzjw** [comment]
Hi @huitema, following `picoquic_format_path_challenge_frame` and `picoquic_decode_path_challenge_frame`, I'm currently debugging the demo program on Mininet. 

Based on my observation, on the client side, when `cnx->cnx_state==picoquic_state_ready`, the client calls `picoquic_format_path_challenge_frame`, with `path_x->local_addr=client_alt_address` which specified in the cmdline, `path_x->peer_addr=server_address`, which is expected.

On the server side, when `picoquic_decode_path_challenge_frame` is called, `addr_from` is the default IP from the client(i.e., the IP used for `path[0]`), rather than `client_alt_address`. So after the client finishes `picoquic_decode_path_response_frame`, it actually verified the default `path[0]`, rather than the newly added alternative path. And the new path is never selected. 

For reference, the Mininet topology I used:

```
10.0.1.2 <-----link1------>      <-----link1------> 10.0.3.2
  HostA                    Router                     HostB
10.0.2.2 <-----link2------>      <-----link2------> 10.0.4.2
```

In this case, picoquicdemo server runs on `HostA`, the client runs on `HostB`, `server_address=10.0.1.2` and `client_alt_address=10.0.4.2`. All the traffic including the probe packets between Router and `HostB` goes through `link1` which is not expected.

<details>
  <summary>FYI, The code for the topology</summary>

```python
from mininet.net import Mininet
from mininet.cli import CLI
from mininet.link import Link, TCLink,Intf
from subprocess import Popen, PIPE
from mininet.log import setLogLevel

if '__main__' == __name__:
  setLogLevel('info')
  net = Mininet(link=TCLink)

  h1 = net.addHost('h1')
  h2 = net.addHost('h2')
  r1 = net.addHost('r1')

  linkopt={'bw':1000}

  net.addLink(r1, h1 ,cls=TCLink, **linkopt)
  net.addLink(r1, h1 ,cls=TCLink, **linkopt)
  net.addLink(r1, h2 ,cls=TCLink, **linkopt)
  net.addLink(r1, h2 ,cls=TCLink, **linkopt)
  net.build()

  r1.cmd("ifconfig r1-eth0 0")
  r1.cmd("ifconfig r1-eth1 0")
  r1.cmd("ifconfig r1-eth2 0")
  r1.cmd("ifconfig r1-eth3 0")
  r1.cmd("echo 1 > /proc/sys/net/ipv4/ip_forward")
  r1.cmd("echo 1 > /proc/sys/net/ipv4/conf/all/proxy_arp")

  r1.cmd("ifconfig r1-eth0 10.0.1.1 netmask 255.255.255.0")
  r1.cmd("ifconfig r1-eth1 10.0.2.1 netmask 255.255.255.0")
  r1.cmd("ifconfig r1-eth2 10.0.3.1 netmask 255.255.255.0")
  r1.cmd("ifconfig r1-eth3 10.0.4.1 netmask 255.255.255.0")

  h1.cmd("ifconfig h1-eth0 0")
  h1.cmd("ifconfig h1-eth1 0")
  h2.cmd("ifconfig h2-eth0 0")
  h2.cmd("ifconfig h2-eth1 0")

  h1.cmd("ifconfig h1-eth0 10.0.1.2 netmask 255.255.255.0")
  h1.cmd("ifconfig h1-eth1 10.0.2.2 netmask 255.255.255.0")

  h2.cmd("ifconfig h2-eth0 10.0.3.2 netmask 255.255.255.0")
  h2.cmd("ifconfig h2-eth1 10.0.4.2 netmask 255.255.255.0")

  h1.cmd("ip rule add from 10.0.1.2 table 1")
  h1.cmd("ip rule add from 10.0.2.2 table 2")
  h1.cmd("ip route add 10.0.1.0/24 dev h1-eth0 scope link table 1")
  h1.cmd("ip route add 10.0.3.0/24 dev h1-eth0 scope link table 1")
  
  h1.cmd("ip route add 10.0.2.0/24 dev h1-eth1 scope link table 2")
  h1.cmd("ip route add 10.0.4.0/24 dev h1-eth1 scope link table 2")
  h1.cmd("ip route add default scope global nexthop via 10.0.1.1 dev h1-eth0")

  h2.cmd("ip rule add from 10.0.3.2 table 1")
  h2.cmd("ip rule add from 10.0.4.2 table 2")
  h2.cmd("ip route add 10.0.3.0/24 dev h2-eth0 scope link table 1")
  h2.cmd("ip route add 10.0.1.0/24 dev h2-eth0 scope link table 1")
  
  h2.cmd("ip route add 10.0.4.0/24 dev h2-eth1 scope link table 2")
  h2.cmd("ip route add 10.0.2.0/24 dev h2-eth1 scope link table 2")
  h2.cmd("ip route add 10.0.1.0/24 dev h2-eth1 scope link table 2")
  h2.cmd("ip route add default scope global nexthop via 10.0.3.1 dev h2-eth0")

  CLI(net)
  net.stop()
```
  </summary>
</details>

I guess this is caused by some misconfigured routes, however, I still cannot figure out the cause. Any suggestions are much appreciated. Thanks.

**Comment by huitema** [comment]
This might be an issue with BSD. If you look at the code in "picosock.c", you will see that the source address handling depends on the BSD specific socket option "IP_SENDSRCADDR". The tests so far have been done with Windows and with Linux. Something is wrong: either the option IP_SENDSRCADDR is not working as expected; or there is a bug in the picoquic code.

The first step is to debug that when the probe is sent to the socket, the source address parameter is correctly passed to the loop, from the loop to `picoquic_sendmsg`, and then to `picoquic_socks_cmsg_format`, and verify that the option is correctly encoded per BSD requirements. That will detect a potential error in picoquic.

**Comment by clarkzjw** [comment]
Thanks for the info. My tests are done in a Debian VM, the host OS is macOS though.

I added some logs on the client side, and I think the source address is correctly passed to `picoquic_socks_cmsg_format` for the probe packet. But that packet was never sent through the alternative interface. 

```
Connection established. Version = 1, I-CID: 2d46520950b8c8c7, verified: 1
=====server address: 10.0.1.2
=====client alt address: 10.0.4.2
quicctx.c:picoquic_create_path, local: 10.0.4.2
quicctx.c:picoquic_create_path, peer: 10.0.1.2
quicctx.c:picoquic_create_path, cnx->nb_paths: 2

quicctx.c:picoquic_probe_new_path_ex, addr_from: 10.0.1.2
quicctx.c:picoquic_probe_new_path_ex, addr_to: 10.0.4.2

sender.c:picoquic_prepare_packet_ex path id: 1, verified: 0, resp required: 0, challenge required: 1, state: 14
sender.c:picoquic_prepare_packet_ready, path_x->local_addr: 10.0.4.2, path_x->peer_addr: 10.0.1.2
sender.c:picoquic_prepare_packet_ex path id: 1, verified: 0, resp required: 0, challenge required: 1, state: 14
picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.4.2 
## I think this would be the src IP for the probe packet

frames.c:picoquic_decode_frames
frames.c:picoquic_decode_path_response_frame, path_x->local_addr: 10.0.3.2

picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.3.2
picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.3.2
picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.3.2
picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.3.2
picosocks.c:picoquic_socks_cmsg_format, str_addr_from: 10.0.3.2
```

**Comment by huitema** [comment]
Interesting. That could be due to the "interface identifier" parameter. Looking at the[ Ubuntu manual page for "ip"](https://manpages.ubuntu.com/manpages/xenial/en/man7/ip.7.html), I see:
```

**Comment by clarkzjw** [comment]
@huitema Thanks for the info! I guess it might be the case. Following your advice, I added some workaround in the code to deal with `ipi_ifindex`, and it seems it finally utilizing both interfaces!

My current commit to test: https://github.com/clarkzjw/picoquic/commit/a4c12cde3823cb54db185fe3acb4666f663e87cc. There are many `printf` for me to debug, but the workaround is here: https://github.com/clarkzjw/picoquic/commit/a4c12cde3823cb54db185fe3acb4666f663e87cc#diff-4c8a54be02d908bcc6d7b7ec707dc1abdc4b1b30665a9417f0f586ea10fe7ac5R609-R619
```c
// TODO
// 10.0.4.2 is on h2-eth1 on the client
// 10.0.1.2 is on h1-eth0 on the server
if (strcmp(str_addr_from, "10.0.4.2") == 0) {
    pktinfo->ipi_ifindex = if_nametoindex("h2-eth1");
    printf("%s:%d, pktinfo->ipi_ifindex = h2-eth1: %d\n", __FUNCTION__, __LINE__, pktinfo->ipi_ifindex);
} else if (strcmp(str_addr_from, "10.0.1.2") == 0) {
    pktinfo->ipi_ifindex = if_nametoindex("h1-eth0");
} else {
    pktinfo->ipi_ifindex = (unsigned long)dest_if;
}
```
As you've mentioned, without these workarounds, for probe packets, `dest_if` passed in is 0. See the client logs below:

```
root@ubuntu:/home/clarkzjw/code/picoquic# ./picoquicdemo -M 2 -A 10.0.4.2 -o /tmp 10.0.1.2 4433 test.bin
Starting Picoquic (v1.01a) connection to server = 10.0.1.2, port = 4433
Set ALPN to h3 based on stored ticket
Set version to 0x00000001 based on stored ticket
Testing scenario: <test.bin>
picoquic_create_path:1381, peer_addr: 10.0.1.2
picoquic_create_path:1452, picoquic_create_path done, cnx->nb_paths: 1

Max stream id bidir remote before start = 0 (0)
Starting client connection. Version = 1, I-CID: 7ee71e87fae8e2ec
Max stream id bidir remote after start = 2048 (2048)
Max stream id bidir remote after 0rtt = 2048 (2048)
Opening stream 0 to GET /test.bin
Waiting for packets.
Client port (AF=2): 44400.
Negotiated ALPN: h3
Almost ready!

picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
Connection established. Version = 1, I-CID: 7ee71e87fae8e2ec, verified: 1
=====server address: 10.0.1.2
=====client alt address: 10.0.4.2
picoquic_create_path:1377, local_addr: 10.0.4.2
picoquic_create_path:1381, peer_addr: 10.0.1.2
picoquic_create_path:1452, picoquic_create_path done, cnx->nb_paths: 2

picoquic_probe_new_path_ex:1947, addr_from: 10.0.1.2
picoquic_probe_new_path_ex:1948, addr_to: 10.0.4.2

picoquic_prepare_packet_ex:4677, path id: 1, verified: 0, resp required: 0, challenge required: 1, state: 14
picoquic_prepare_packet_ready:3825, path_x->local_addr: 10.0.4.2
picoquic_prepare_packet_ready:3826, path_x->peer_addr: 10.0.1.2
picoquic_prepare_packet_ex:4677, path id: 1, verified: 0, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=0
# I think this is when the probe packets are about to send





picoquic_create_path:1377, local_addr: 10.0.4.2
picoquic_create_path:1381, peer_addr: 10.0.1.2
picoquic_create_path:1452, picoquic_create_path done, cnx->nb_paths: 3
# Although I'm not sure why a third path is created
# In the following data transmission, path 0 and 3 are being utilized




picoquic_decode_frames:5113, calling picoquic_decode_path_challenge_frame: path verified: 0, response required: 0
picoquic_decode_path_challenge_frame:4153, path_x->response_required = 1
picoquic_decode_path_challenge_frame:4154, addr_from: 10.0.1.2
picoquic_decode_path_challenge_frame:4155, addr_to: 10.0.4.2

picoquic_decode_frames:5117, after picoquic_decode_path_challenge_frame: path verified: 0, response required: 1
/home/clarkzjw/code/picoquic/picoquic/frames.c:picoquic_decode_frames
picoquic_decode_frames:5124, calling picoquic_decode_path_response_frame: path verified: 0, response required: 1
picoquic_decode_path_response_frame:4191, path_x->local_addr: 10.0.4.2
picoquic_decode_path_response_frame:4192, path_x->peer_addr: 10.0.1.2
picoquic_decode_frames:5129, after picoquic_decode_path_response_frame: path verified: 0, response required: 1
picoquic_select_next_path_mp:4374, path 2 response_required
picoquic_prepare_packet_ex:4677, path id: 2, verified: 0, resp required: 1, challenge required: 1, state: 14
picoquic_prepare_packet_ready:3825, path_x->local_addr: 10.0.4.2
picoquic_prepare_packet_ready:3826, path_x->peer_addr: 10.0.1.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 0, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
/home/clarkzjw/code/picoquic/picoquic/frames.c:picoquic_decode_frames
picoquic_decode_frames:5124, calling picoquic_decode_path_response_frame: path verified: 0, response required: 0
picoquic_decode_path_response_frame:4191, path_x->local_addr: 10.0.4.2
picoquic_decode_path_response_frame:4192, path_x->peer_addr: 10.0.1.2
picoquic_decode_path_response_frame:4197, found_challenge = 1;
picoquic_decode_path_response_frame:4206, challenge_verified = 1
picoquic_decode_frames:5129, after picoquic_decode_path_response_frame: path verified: 1, response required: 0
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
......
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 14
picoquic_socks_cmsg_format:608, addr_from: 10.0.4.2
picoquic_socks_cmsg_format:614, pktinfo->ipi_ifindex is set to h2-eth1: 3, passed in dest_if=3
Stream 0 ended after 2343608 bytes
All done, Closing the connection.
picoquic_socks_cmsg_format:608, addr_from: 10.0.3.2
picoquic_prepare_packet_ex:4677, path id: 2, verified: 1, resp required: 0, challenge required: 1, state: 17
Received a request to close the connection.
The connection is closed!
Out of 1 zero RTT packets, 0 were acked by the server.
Quic Bit was greased by the client.
Quic Bit was greased by the server.
ECN was not received.
ECN was acknowledged (ect0: 88, ect1: 0, ce: 0).
Received 2343642 bytes in 0.171470 seconds, 109.343535 Mbps.
max_data_local: 10176552
max_max_stream_data_local: 7034460
max_data_remote: 1048576
max_max_stream_data_remote: 2048
ack_delay_remote: 1000 ... 1000
max_ack_gap_remote: 15
ack_delay_local: 1000 ... 25000
max_ack_gap_local: 4
max_mtu_sent: 1252
max_mtu_received: 1440
Received ticket from 10.0.1.2 (h3):
ticket time = 1654415377243, kx = 17, suite = 1301, 114 ticket, 32 secret.
lifetime = 100000, age_add = 3aa6f6c6, 0 nonce, 93 ticket, 8 extensions.
ticket extensions: 42(ED: ffffffff),
Client exit with code = 0
```

Now I can observe data transmission traffic on both interfaces on the client, thus multipath is being utilized. So the problem might be related to `ipi_ifindex` affecting probe packets. For reference, the routing table on the client is attached below, and the topology is still the same like https://github.com/private-octopus/picoquic/issues/1134#issuecomment-1146461513:
```
root@ubuntu:/home/clarkzjw/code/picoquic# ip rule show
0:      from all lookup local
32764:  from 10.0.4.2 lookup 2
32765:  from 10.0.3.2 lookup 1
32766:  from all lookup main
32767:  from all lookup default
root@ubuntu:/home/clarkzjw/code/picoquic# ip route show table 1
10.0.1.0/24 dev h2-eth0 scope link 
10.0.3.0/24 dev h2-eth0 scope link 
root@ubuntu:/home/clarkzjw/code/picoquic# ip route show table 2
10.0.2.0/24 dev h2-eth1 scope link 
10.0.4.0/24 dev h2-eth1 scope link
```

**Comment by huitema** [comment]
The specification says that 0 is supposed to be a "wildcard" for ipi_ifindex. I wonder why this is not working. Could it be because mininet creates an interface with index 0? In any case, we need to do the real fix, which is to add an extra "if_index" parameter to
```
int picoquic_probe_new_path(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
    const struct sockaddr* addr_to, uint64_t current_time);
```
I don't want to change that API, but instead to create an extended version with an additional parameter. This is a bit complicated because we already have a function:
```
int picoquic_probe_new_path_ex(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
    const struct sockaddr* addr_to, uint64_t current_time, int to_preferred_address); 
```
However, that function is only exposed in the "private" header, so modifying it will not break implementations that followed the rules. My proposal:

1) update `picoquic_probe_new_path_ex` to:
```
int picoquic_probe_new_path_ex(picoquic_cnx_t* cnx, const struct sockaddr* addr_from,
    const struct sockaddr* addr_to, int if_index, uint64_t current_time, int to_preferred_address); 
```
 2) fix the existing calls to `picoquic_probe_new_path_ex` with setting `if_index=0`.
 
 3) move the definition of `picoquic_probe_new_path_ex` to `picoquic.h`
 
 4) update the code in `picoquicdemo.c` to call `picoquic_probe_new_path_ex`, somehow adding an if_index parameter in the option "-A".
 
 I could do that, but I cannot test it. Do you want to do it?

**Comment by huitema** [comment]
There may be something simpler. Your configuration sets the server with two addresses:
```
10.0.1.2 <-----link1------>      <-----link1------> 10.0.3.2
  HostA                    Router                     HostB
10.0.2.2 <-----link2------>      <-----link2------> 10.0.4.2
```
But the probe API is designed to only specify a single address for the server. This is because most services today use server farms, so specifying a different address risks sending the packet to a different server, maybe in some replicated location. There are two solutions:

1) Change your configuration so the server has only one address, and one link to the router. That will simulate the most common behavior. That implies setting 2 routes to the server address, one for each interface.

2) Change the calls in picoquicdemo so they specify both the source and the destination address, but that's going to be complicated.

**Comment by clarkzjw** [comment]
```
Could it be because mininet creates an interface with index 0?
```
I tested `if_indextoname(unsigned int ifindex, char *ifname)` in Mininet, passing `ifindex=0`, and it returns NULL, which I guess is the expected behaviour?
```
root@ubuntu:/home/clarkzjw/code/picoquic# ip link 
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: h2-eth0@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP mode DEFAULT group default qlen 1000
    link/ether 16:b8:8a:31:ee:0d brd ff:ff:ff:ff:ff:ff link-netnsid 0
3: h2-eth1@if5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP mode DEFAULT group default qlen 1000
    link/ether c2:51:ba:03:bb:0c brd ff:ff:ff:ff:ff:ff link-netnsid 0
```


```
1. Change your configuration so the server has only one address, and one link to the router. That will simulate the most common behavior. That implies setting 2 routes to the server address, one for each interface.

2. Change the calls in picoquicdemo so they specify both the source and the destination address, but that's going to be complicated.
```
I think maybe we can go for option 1 first, fixing the issue by your proposal in https://github.com/private-octopus/picoquic/issues/1134#issuecomment-1146838889. After all, this is the most common topology of multi-homed users today. My workaround above actually falls into this as well, despite my initial topology being two links on each side.

**Comment by huitema** [comment]
Seems like a good plan!

**Comment by clarkzjw** [comment]
😄 Oh sorry, I missed the last sentence in https://github.com/private-octopus/picoquic/issues/1134#issuecomment-1146838889. Yeah, I think maybe I can do that. But I cannot start working on it and submit a PR until Tuesday if you don't mind.

**Comment by huitema** [comment]
No problem. I am quite busy, won't be able to do that sooner anyhow...

**Comment by huitema** [comment]
Fixed in PR #1369
